In [1]:
from pymongo import MongoClient

In [2]:
client = MongoClient("mongodb://10.10.10.46:20000/")

In [3]:
db = client.demo_docs

In [62]:
doc_info = db.doc_info

In [4]:
patient_basic_info = db.patient_basic_info

In [5]:
demo_xinshuai = client.demo_xinshuai_201803241749

In [6]:
patient_medical_history = demo_xinshuai['patient_medical_history']

In [7]:
filter = {'toReport':'是','$or':[{'_delete':{'$exists':False}},{'_delete':False}]}

In [8]:
patient_medical_history.count_documents(filter)
# 上报患者数目

119014

In [15]:
import pandas as pd

In [61]:
# 所以上报患者ID
patient_df = pd.DataFrame(list(patient_medical_history.find(filter,{'_id':0,'patientId':1})))
patient_df.head()

patientId  hospitalId
0  23f47cb1-efaf-11e7-ae39-53ece1750c11         NaN
1  292eaed0-efaf-11e7-ae39-53ece1750c11         NaN
2  2fadbd01-efaf-11e7-8fb4-0ba8851511e3         NaN
3  5c456a41-c46a-11e7-93ca-d39b9bc884de         NaN
4  4c6c8f91-c46a-11e7-8534-c51051c27839         NaN

In [40]:
patient_df.count()

patientId    119014
dtype: int64

In [38]:
patient_df.to_csv('心衰上报患者ID.csv')

In [26]:
patient_df.iloc[:,0]

0         23f47cb1-efaf-11e7-ae39-53ece1750c11
1         292eaed0-efaf-11e7-ae39-53ece1750c11
2         2fadbd01-efaf-11e7-8fb4-0ba8851511e3
3         5c456a41-c46a-11e7-93ca-d39b9bc884de
4         4c6c8f91-c46a-11e7-8534-c51051c27839
                          ...                 
119009        cd403ebd8bf15e0c1dfef66a1d766068
119010        3341039269c7d350563f376e22566c25
119011        9385768f18a7918c095383cb1e2239a9
119012        4f90190d8c82f29aa019dbb075016e10
119013        4164b5a89cdc271bf17936fc933e28bd
Name: patientId, Length: 119014, dtype: object

In [32]:
patient_filter = {'patientId':{'$in':patient_df.iloc[:,0].values.tolist()}}

In [44]:
patient_pd = pd.DataFrame(list(patient_basic_info.find(patient_filter,{'patientId':1,'patientBasicInfoIdCard':1,'patientBasicInfoPatientUserName':1,'patientBasicInfoGender':1,'patientBasicInfoBirthday':1})),dtype=str)
patient_pd.head()

_id   patientBasicInfoPatientUserName  \
0      00009f2317d433c50aea0c6840327e52  df845cde3f03d547a651550ef24a1761   
1      0000fbc3bc845017df1c05471abc7445  3456e3e823376d3cdfad3dcd6eea373a   
2      0001491d8a76be51aafcd0ae94250098  026a85fa600b04b6230ed82ec910117e   
3  00018741-f773-11e8-8a19-e1b86dde47c7  4d5275b6088baa974cf63891ab6c6ec6   
4      00024a007dc0dad6a19e990ed1fe9e22  1cc45584c6d15bca0011556f56b704bc   

             patientBasicInfoIdCard patientBasicInfoGender  \
0  f67b341df6ad415c1dd13da46ba8579a                      男   
1  bfde2147bc725dd40762109bec1eba13                      男   
2  11ea8159cb60cc5b0b149ac83f256640                      男   
3  a379b2c8175e78b93a58a4dc91cbf88b                      男   
4  4b593047923b26a7dfc1ff629364d35c                      女   

  patientBasicInfoBirthday                             patientId  
0          -820396800000.0      00009f2317d433c50aea0c6840327e52  
1          -750240000000.0      0000fbc3bc845017df1c05471abc7445  
2          -354614400000.0      0001491d8a76be51aafcd0ae94250098  
3          -378115200000.0  00018741-f773-11e8-8a19-e1b86dde47c7  
4          -981187200000.0      00024a007dc0dad6a19e990ed1fe9e22

In [73]:
doc_filter = dict(patient_filter)
doc_filter['diseaseId'] = 47
doc_filter['$or'] = [{'_delete':{'$exists':False}},{'_delete':False}]
docinfo_df = pd.DataFrame(list(doc_info.find(doc_filter)))
docinfo_df.head()

_id                      oldHistoryId  \
0  55ab2cd1-1024-11ea-a7d9-d1a8c1532bf1  72fdec75018711e99d03509a4c82024b   
1  0ffc7580-86a5-11e9-affe-8d50ca0b1a6b  87388a71ef4111e8b9c4509a4c868a0c   
2  267f3ef1-86a5-11e9-affe-8d50ca0b1a6b  9127ac2811d411e9b99a509a4c868a0c   
3  8177b4f1-32ca-11ea-a691-d9abf2838a50  6949cb1efbd711e98f28509a4c82024b   
4  3959aee1-1023-11ea-9e0b-330dfb01131e  014d68ae9e7e11e98e63509a4c82024b   

                    globalPatientId                   globalHistoryId  \
0  00009f2317d433c50aea0c6840327e52  d7693d04bb65e9cfa3c50891f76ca204   
1  0000fbc3bc845017df1c05471abc7445  b2748d869dd7bc8bf1f1001e2960bca1   
2  0000fbc3bc845017df1c05471abc7445  2d07ed4d6f95218256f5c43f9a868e4a   
3  ca3630e8b8854bc13994a74680ece105  d92054fa6dc0a31793e9b254abb6e4d0   
4  0001491d8a76be51aafcd0ae94250098  771651bec615e2c276b0e04e12901d10   

   diseaseId  hospitalId  deptId                             historyId  \
0         47         262       1  0737bf41-b2a6-11e9-bfc1-1d5141e92ecf   
1         47         618       1      b2748d869dd7bc8bf1f1001e2960bca1   
2         47         618       1      2d07ed4d6f95218256f5c43f9a868e4a   
3         47         262       1  e0069240-32c3-11ea-a691-d9abf2838a50   
4         47         262       1  e369f1f1-b2a5-11e9-bfc1-1d5141e92ecf   

  sourceCode                          sourceId  ...   diseaseName  \
0       tbxt  72fdec75018711e99d03509a4c82024b  ...  中国心衰中心数据填报平台   
1        etl  87388a71ef4111e8b9c4509a4c868a0c  ...  中国心衰中心数据填报平台   
2        etl  9127ac2811d411e9b99a509a4c868a0c  ...  中国心衰中心数据填报平台   
3       tbxt  6949cb1efbd711e98f28509a4c82024b  ...  中国心衰中心数据填报平台   
4       tbxt  014d68ae9e7e11e98e63509a4c82024b  ...  中国心衰中心数据填报平台   

                       oldPatientId cur_es  diseaseCode maxs  _verify  \
0                               NaN    NaN          NaN  NaN      NaN   
1  87323ebbef4111e8b9c4509a4c868a0c    NaN          NaN  NaN      NaN   
2  87323ebbef4111e8b9c4509a4c868a0c    NaN          NaN  NaN      NaN   
3  692601a8fbd711e98f28509a4c82024b    NaN          NaN  NaN      NaN   
4                               NaN    NaN          NaN  NaN      NaN   

   patientBasicInfoTelNumber _delete  rc4 doctorId  
0                        NaN     NaN  NaN      NaN  
1                        NaN     NaN  NaN      NaN  
2                        NaN     NaN  NaN      NaN  
3                        NaN     NaN  NaN      NaN  
4                        NaN     NaN  NaN      NaN  

[5 rows x 39 columns]

In [45]:
patient_pd.count()

_id                                93293
patientBasicInfoPatientUserName    93293
patientBasicInfoIdCard             89929
patientBasicInfoGender             93288
patientBasicInfoBirthday           90537
patientId                          93293
dtype: int64

In [47]:
patient_pd.to_csv('心衰上报患者.csv')

In [50]:
# 包含 姓名 性别 出生日期的
patient_pd.loc[(patient_pd.patientBasicInfoPatientUserName.notnull()) & (patient_pd.patientBasicInfoGender.notnull()) & (patient_pd.patientBasicInfoBirthday.notnull())].count()

_id                                90534
patientBasicInfoPatientUserName    90534
patientBasicInfoIdCard             87172
patientBasicInfoGender             90534
patientBasicInfoBirthday           90534
patientId                          90534
dtype: int64

In [60]:
patient_pd.loc[(patient_pd.patientBasicInfoIdCard.isnull()) & (patient_pd.patientBasicInfoPatientUserName.notnull()) & (patient_pd.patientBasicInfoGender.notnull()) & (patient_pd.patientBasicInfoBirthday.notnull())].count()

_id                                3362
patientBasicInfoPatientUserName    3362
patientBasicInfoIdCard                0
patientBasicInfoGender             3362
patientBasicInfoBirthday           3362
patientId                          3362
dtype: int64

In [51]:
# 排除IDcard不正确的

In [52]:
import pymysql

In [53]:
conn = pymysql.connect(host='10.10.10.32', port=3306, user='root', password='root', db='platform_patientindex', charset='utf8',cursorclass=pymysql.cursors.DictCursor)

In [55]:
fix_idcard_df = pd.read_sql_query("select * from index_failover_center", conn)

In [56]:
fix_idcard_df.head()

id         create_time         update_time  \
0  00441aabaca2bcecf8870680ea529037 2019-12-26 18:25:52 2019-12-26 18:25:52   
1  00cf83a98f46fcf7066ed01f72a562f5 2019-12-26 18:25:52 2019-12-26 18:25:52   
2  0129210c45d40afba018329aa66ceed6 2019-12-26 18:25:52 2019-12-26 18:25:52   
3  0195b53711609d4ae26b0bcbe713fa28 2019-12-26 18:25:52 2019-12-26 18:25:52   
4  0220bead89567f2d67506e9bf61426b7 2019-12-26 18:25:52 2019-12-26 18:25:52   

                        crypto_text plain_text  
0  06409663226af2f3114485aa4e0a23b4        158  
1  1eac12f6683635318966bdff19deb7df         未有  
2  56521b4d2824c0cb09bb240f8461acad     326598  
3  fb21b421091fdad755f4eba24d8f8946   56655656  
4  e3ceb5881a0a1fdaad01296d7554868d     222222

In [58]:
# 不在idcard中的
patient_pd.loc[-patient_pd.patientBasicInfoIdCard.isin(fix_idcard_df.crypto_text.values.tolist())].count()

_id                                91636
patientBasicInfoPatientUserName    91636
patientBasicInfoIdCard             88272
patientBasicInfoGender             91631
patientBasicInfoBirthday           88889
patientId                          91636
dtype: int64

In [67]:
# 没有身份中号的患者
noidcard_df = patient_pd.loc[patient_pd.patientBasicInfoIdCard.isnull()|patient_pd.patientBasicInfoIdCard.isin(fix_idcard_df.crypto_text.values.tolist())]
noidcard_df.head()

_id   patientBasicInfoPatientUserName  \
2       0001491d8a76be51aafcd0ae94250098  026a85fa600b04b6230ed82ec910117e   
58  00288531-f151-11e7-866f-b9d01e432bdb  e8dc1577ae4368d63bbd83f337ec86db   
67      002d9678bd490f03d3f03f2669542ace  b097da8688606998005bc572998fd2d5   
69  002de2d1-4126-11e8-8cfe-8fb33112142e  1e30ac70ec293fcb7453ef627f94fa16   
82      00351256aa360150c99634ba5131ca22  9a6c3f87a6b847de65cfc1edbc0bb871   

              patientBasicInfoIdCard patientBasicInfoGender  \
2   11ea8159cb60cc5b0b149ac83f256640                      男   
58                               NaN                      女   
67  a6b008e69f9ad0d547837686f6b1f1cd                      男   
69  11ea8159cb60cc5b0b149ac83f256640                      女   
82  a6b008e69f9ad0d547837686f6b1f1cd                      男   

   patientBasicInfoBirthday                             patientId  
2           -354614400000.0      0001491d8a76be51aafcd0ae94250098  
58         -1365148800000.0  00288531-f151-11e7-866f-b9d01e432bdb  
67          -328176000000.0      002d9678bd490f03d3f03f2669542ace  
69          -831196800000.0  002de2d1-4126-11e8-8cfe-8fb33112142e  
82          -492681600000.0      00351256aa360150c99634ba5131ca22

In [69]:
noidcard_df.count()

_id                                5021
patientBasicInfoPatientUserName    5021
patientBasicInfoIdCard             1657
patientBasicInfoGender             5021
patientBasicInfoBirthday           5010
patientId                          5021
dtype: int64

In [76]:
# 没有身份证号的按医院分布
noidcardhospital_df = docinfo_df.loc[docinfo_df.patientId.isin(noidcard_df.patientId.values.tolist())]

In [104]:
hospital_g = noidcardhospital_df[['hospitalId','patientId']].groupby('hospitalId').count().apply(lambda x:x.sort_values(ascending=False)).reset_index()

In [121]:
hospital_df = docinfo_df[['hospitalId','patientId']].groupby('hospitalId').count().apply(lambda x:x.sort_values(ascending=False)).reset_index()
hospital_df.head()

hospitalId  patientId
0         262       4396
1         346       4250
2         140       4019
3         677       3159
4         323       3066

In [105]:
hospital_g

hospitalId  patientId
0           17       2001
1          290       1249
2          730        474
3          346        384
4          696        376
..         ...        ...
69         217          1
70         723          1
71         276          1
72         715          1
73         211          1

[74 rows x 2 columns]

## 结论
上报患者总数：93292  
有身份证的：88272  
姓名+性别+出生日期：90534  
没有身份证号有 姓名+性别+出生日期：3362

没有身份证号：5021

In [106]:
hospital_g.to_csv('没有身份证号医院分布.csv')

In [111]:
hospital_g.head()

hospitalId  patientId
0          17       2001
1         290       1249
2         730        474
3         346        384
4         696        376

In [107]:
conn2 = pymysql.connect(host='10.10.10.13', port=3306, user='root', password='root', db='emdata', charset='utf8',cursorclass=pymysql.cursors.DictCursor)

In [110]:
hospitalinfo_df = pd.read_sql_query("select id,name from pr_hospital_info", conn2)
hospitalinfo_df.head()

id     name
0   2     协和医院
1   3    待审核医院
2   7  太原市中心医院
3   8    山西大医院
4   9   中日友好医院

In [115]:
nohospitalinfo_df = hospitalinfo_df.loc[hospitalinfo_df.id.isin(hospital_g.hospitalId.values.tolist())]
nohospitalinfo_df.head()

id                   name
2    7                太原市中心医院
8   15  南京医科大学第一附属医院(江苏省人民医院)
10  17           重庆医科大学附属第一医院
18  25             苏州大学附属第二医院
21  28              无锡市第三人民医院

In [119]:
hospital_gn = hospital_g.join(nohospitalinfo_df.set_index('id'), on='hospitalId')
hospital_gn.head()

hospitalId  patientId                  name
0          17       2001          重庆医科大学附属第一医院
1         290       1249  吉林大学第三医院（吉林大学中日联谊医院）
2         730        474              哈尔滨市第一医院
3         346        384     武汉大学人民医院（湖北省人民医院）
4         696        376          山西医科大学附属第一医院

In [120]:
hospital_gn.to_csv('没有身份证号医院分布.csv')

In [128]:
agn = hospital_gn.join(hospital_df.set_index('hospitalId'), lsuffix='_没有身份证', rsuffix='_总数', on='hospitalId')
agn.head()

hospitalId  patientId_没有身份证                  name  patientId_总数
0          17             2001          重庆医科大学附属第一医院          2037
1         290             1249  吉林大学第三医院（吉林大学中日联谊医院）          1296
2         730              474              哈尔滨市第一医院          1572
3         346              384     武汉大学人民医院（湖北省人民医院）          4250
4         696              376          山西医科大学附属第一医院           695

In [129]:
agn.to_csv('没有身份证号医院分布.csv')